rust/bindings/build.rs

```diff
@@ -0,0 +1,260 @@
+                    .into_iter()
+                    .map(|cd| space.add_type(&cd.schema).map(|id| (cd.name, id)))
+                    .collect::<Result<Vec<_>, _>>()?,
+            })
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let method_code = method_ir.iter().try_fold(
+        TokenStream::new(),
+        |mut acc,
+         Method {
+             name,
+             result,
+             params,
+         }| {
+            let fn_name = syn::parse_str::<Ident>(&name.replace('.', "_"))?;
+            let ret = result
+                .as_ref()
+                .map(|id| space.get_type(id).map(|ty| ty.ident()))
+                .transpose()?;
+            let vars = params
+                .iter()
+                .map(|(name, _)| syn::parse_str::<Ident>(name))
+                .collect::<Result<Vec<_>, _>>()?;
+            let params = params
+                .iter()
+                .map(|(name, id)| {
+                    let name = syn::parse_str::<Ident>(name)?;
+                    let ty = space.get_type(id)?.parameter_ident();
+                    anyhow::Ok(quote! { #name: #ty })
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+            match ret {
+                Some(ret) => acc.extend(quote! {
+                    fn #fn_name(&mut self, #(#params),*) -> Result<#ret, Self::Error> {
+                        self.call(#name, (#(#vars,)*))
+                    }
+                }),
+                None => bail!("notifications are not currently supported"),
+            };
+            anyhow::Ok(acc)
+        },
+    )?;
+
+    Ok(quote! {
+        #![allow(clippy::to_string_trait_impl, clippy::clone_on_copy)]
+        use serde::{Serialize, Deserialize, de::DeserializeOwned};
+
+        #[allow(non_snake_case, unused)]
+        pub trait Api {
+            type Error;
+            fn call<T: DeserializeOwned>(
+                &mut self,
+                method: impl Into<String>,
+                params: impl ez_jsonrpc::params::SerializePositional
+            ) -> Result<T, Self::Error>;
+            #method_code
+        }
+
+        #space
+    })
+}
+
+struct Method {
+    name: String,
+    result: Option<typify::TypeId>,
+    params: Vec<(String, typify::TypeId)>,
+}
+
+fn extract(s: &str) -> anyhow::Result<&str> {
+    let (_whole, grp) = lazy_regex::regex_captures!("^#/components/schemas/(.*)", s)
+        .context(format!("unsupported reference format in `{}`", s))?;
+    Ok(grp)
+}
+
+/// OpenRPC uses `#/components/schemas/Foo`,
+/// but JSON Schema (and [typify]) uses `#/definitions/Foo`.
+fn rewrite_references(
+    schema: &mut Schema,
+    mut rewrite: impl FnMut(&mut String) -> anyhow::Result<()>,
+) -> anyhow::Result<()> {
+    if let Schema::Object(SchemaObject {
+        reference,
+        subschemas,
+        array,
+        object,
+        ..
+    }) = schema
+    {
+        reference.as_mut().map(cast(&mut rewrite)).transpose()?;
+
+        for child in array
```

Comment:
This loop is especially painful to read. It would really help if at least it's major blocks were documented. 

E.g.

```rust
// Does that
 .flat_map(
                |ArrayValidation {
                     items,
                     additional_items,
                     contains,
                     ..
                 }| {
                    [additional_items, contains]
                        .into_iter()
                        .flat_map(Option::as_deref_mut)
                        .chain(items.iter_mut().flat_map(|it| match it {
                            SingleOrVec::Vec(it) => &mut it[..],
                            SingleOrVec::Single(it) => std::slice::from_mut(&mut **it),
                        }))
                },
            )
// Does this
            .chain(object.as_deref_mut().iter_mut().flat_map(
                |ObjectValidation {
                     properties,
                     pattern_properties,
                     additional_properties,
                     property_names,
                     ..
                 }| {
                    [additional_properties, property_names]
                        .into_iter()
                        .flat_map(Option::as_deref_mut)
                        .chain(properties.values_mut())
                        .chain(pattern_properties.values_mut())
                },
            ))
```

Reply:
I don't know what comments I could add that aren't redundant with the code - do you have any suggestions?
```diff,rust
            .flat_map(
+               // comb any schemas in the "array" member
                |ArrayValidation {
                     items,
                     additional_items,
                     contains,
                     ..
                 }| {
                    [additional_items, contains]
                        .into_iter()
                        .flat_map(Option::as_deref_mut)
                        .chain(items.iter_mut().flat_map(|it| match it {
                            SingleOrVec::Vec(it) => &mut it[..],
                            SingleOrVec::Single(it) => std::slice::from_mut(&mut **it),
                        }))
                },
            )
            .chain(object.as_deref_mut().iter_mut().flat_map(
+               // comb any schemas in the "object" member
                |ObjectValidation {
                     properties,
                     pattern_properties,
                     additional_properties,
                     property_names,
                     ..
                 }| {
                    [additional_properties, property_names]
                        .into_iter()
                        .flat_map(Option::as_deref_mut)
                        .chain(properties.values_mut())
                        .chain(pattern_properties.values_mut())
                },
            ))
            .chain(subschemas.as_deref_mut().into_iter().flat_map(
+               // comb any schemas in the "anyOf"/"oneOf" etc members
                |SubschemaValidation {
                     all_of,
                     any_of,
                     one_of,
                     not,
                     if_schema,
                     then_schema,
                     else_schema,
                 }| {
                    [not, if_schema, then_schema, else_schema]
                        .into_iter()
                        .flat_map(Option::as_deref_mut)
                        .chain([all_of, any_of, one_of].into_iter().flatten().flatten())
                },
            ))
```

Reply:
The above is good enough for my taste, just something that gives an idea to the reader without having to understand what `ArrayValidation, ObjectValidation and SubschemaValidation` are. When familiar with the subject and the libraries used it does become redundant, but that can be said for a lot of other code out there. Having some clues around makes the reader feel welcome and hopefully eager to contribute. 

I'm often reviewing code without opening it in my IDE, which means I can't navigate through struct definitions and such. This gives an interesting perspective on readability. I hope it's not too annoying.